src/vorbis_decoder.cc

```diff
@@ -246,23 +246,24 @@ int VorbisDecoder::process_single_fr(Encoder* encoder, Buffer* buffer) {
         for (int channel = 0; channel < vi->channels; ++channel) {
             encode_buffer[channel] = new int32_t[samples_per_channel];
         }
-
-        for (int i = 0; i < samples_per_channel; ++i) {
-
+        
+        long sample = 0;
+        for (long i = 0; i < samples_per_channel; ++i) {
             for (int channel = 0; channel < vi->channels; ++channel) {
                 switch (word) {
 
-                    case sizeof(int32_t):
-                        encode_buffer[channel][i] = decode_buffer.as_int[i];
-                        break;
+                case sizeof(int32_t):
+                    encode_buffer[channel][i] = decode_buffer.as_int[sample];
+                    break;
 
-                    case sizeof(short):
-                        encode_buffer[channel][i] = decode_buffer.as_short[i];
-                        break;
+                case sizeof(short):
+                    encode_buffer[channel][i] = (int32_t)decode_buffer.as_short[sample];
```

Comment:
In place of the extra `sample` variable, use `i*vi->channels + channel`. Also, I think the cast here is extraneous as it should be done implicitly by the compiler.